### ISSUE TITLE
Array Slicing Error

### ISSUE THREAD

**Issue by alanhdu** [numba/numba#1312]
I needed to replicate `np.diff`'s behavior when I ran into some problems. If I'm doing something horribly wrong, consider this a request to implement `np.diff` :wink: .

Miniconda w/ Python 3.4:
```Python
In [1]: import numba
In [2]: import numpy as np
In [3]: numba.__version__
Out[3]: '0.20.0'

In [4]: @numba.jit(nopython=True)
   ...: def f():
   ...:     x = np.arange(5)
   ...:     return x[1:] - x[:-1]    # something like `x[1:].sum()` works fine though
   ...:     
In [5]: f()
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/lowering.py in lower_block(self, block)
    171             try:
--> 172                 self.lower_inst(inst)
    173             except LoweringError:

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/lowering.py in lower_inst(self, inst)
    211             ty = self.typeof(inst.target.name)
--> 212             val = self.lower_assign(ty, inst)
    213             self.storevar(val, inst.target.name)

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/lowering.py in lower_assign(self, ty, inst)
    360         elif isinstance(value, ir.Expr):
--> 361             return self.lower_expr(ty, value)
    362 

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/lowering.py in lower_expr(self, resty, expr)
    541         if expr.op == 'binop':
--> 542             return self.lower_binop(resty, expr)
    543         elif expr.op == 'inplace_binop':

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/lowering.py in lower_binop(self, resty, expr)
    405         rhs = self.context.cast(self.builder, rhs, rty, signature.args[1])
--> 406         res = impl(self.builder, (lhs, rhs))
    407         return self.context.cast(self.builder, res, signature.return_type,

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/targets/base.py in __call__(self, builder, args)
   1089     def __call__(self, builder, args):
-> 1090         return self._imp(self._context, builder, self._sig, args)
   1091 

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/targets/npyimpl.py in lower_binary_operator(context, builder, sig, args)
    440         return numpy_ufunc_kernel(context, builder, sig, args, kernel,
--> 441                                   explicit_output=False)
    442     _any = types.Any

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/targets/npyimpl.py in numpy_ufunc_kernel(context, builder, sig, args, kernel_class, explicit_output)
    282         if isinstance(ret_ty, types.Array):
--> 283             output = _build_array(context, builder, ret_ty, arguments)
    284         else:

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/targets/npyimpl.py in _build_array(context, builder, array_ty, arg_arrays)
    258     array_val = arrayobj._empty_nd_impl(context, builder, array_ty,
--> 259                                         dest_shape_tup)
    260     return _prepare_argument(context, builder, array_val._getvalue(), array_ty,

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/targets/arrayobj.py in _empty_nd_impl(context, builder, arrtype, shapes)
   1642             "Don't know how to allocate array with layout '{0}'.".format(
-> 1643                 arrtype.layout))
   1644 

NotImplementedError: Don't know how to allocate array with layout 'A'.

During handling of the above exception, another exception occurred:

LoweringError                             Traceback (most recent call last)
<ipython-input-7-0ec059b9bfe1> in <module>()
----> 1 f()

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    157         assert not kws
    158         sig = tuple([self.typeof_pyval(a) for a in args])
--> 159         return self.compile(sig)
    160 
    161     def inspect_llvm(self, signature=None):

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/dispatcher.py in compile(self, sig)
    320                                           self.py_func,
    321                                           args=args, return_type=return_type,
--> 322                                           flags=flags, locals=self.locals)
    323 
    324             # Check typing error if object mode is used

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in compile_extra(typingctx, targetctx, func, args, return_type, flags, locals, library)
    592     pipeline = Pipeline(typingctx, targetctx, library,
    593                         args, return_type, flags, locals)
--> 594     return pipeline.compile_extra(func)
    595 
    596 

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in compile_extra(self, func)
    315                 raise e
    316 
--> 317         return self.compile_bytecode(bc, func_attr=self.func_attr)
    318 
    319     def compile_bytecode(self, bc, lifted=(), lifted_from=None,

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in compile_bytecode(self, bc, lifted, lifted_from, func_attr)
    324         self.lifted_from = lifted_from
    325         self.func_attr = func_attr
--> 326         return self._compile_bytecode()
    327 
    328     def compile_internal(self, bc, func_attr=DEFAULT_FUNCTION_ATTRIBUTES):

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in _compile_bytecode(self)
    579 
    580         pm.finalize()
--> 581         return pm.run(self.status)
    582 
    583 

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in run(self, status)
    207                     # No more fallback pipelines?
    208                     if is_final_pipeline:
--> 209                         raise patched_exception
    210                     # Go to next fallback pipeline
    211                     else:

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in run(self, status)
    199             for stage, stage_name in self.pipeline_stages[pipeline_name]:
    200                 try:
--> 201                     res = stage()
    202                 except _EarlyPipelineCompletion as e:
    203                     return e.result

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in stage_nopython_backend(self)
    535         """
    536         lowerfn = self.backend_nopython_mode
--> 537         return self._backend(lowerfn, objectmode=False)
    538 
    539     def stage_compile_interp_mode(self):

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in _backend(self, lowerfn, objectmode)
    490             codegen = self.targetctx.jit_codegen()
    491             self.library = codegen.create_library(self.bc.func_qualname)
--> 492         lowered = lowerfn()
    493         signature = typing.signature(self.return_type, *self.args)
    494         cr = compile_result(typing_context=self.typingctx,

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in backend_nopython_mode(self)
    481                 self.return_type,
    482                 self.calltypes,
--> 483                 self.flags)
    484 
    485     def _backend(self, lowerfn, objectmode):

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in native_lowering_stage(targetctx, library, interp, typemap, restype, calltypes, flags)
    726 
    727     lower = lowering.Lower(targetctx, library, fndesc, interp)
--> 728     lower.lower()
    729     if not flags.no_cpython_wrapper:
    730         lower.create_cpython_wrapper(flags.release_gil)

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/lowering.py in lower(self)
     99         if self.generator_info is None:
    100             self.genlower = None
--> 101             self.lower_normal_function(self.fndesc)
    102         else:
    103             self.genlower = self.GeneratorLower(self)

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/lowering.py in lower_normal_function(self, fndesc)
    131         # Init argument values
    132         self.extract_function_arguments()
--> 133         entry_block_tail = self.lower_function_body()
    134 
    135         # Close tail of entry block

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/lowering.py in lower_function_body(self)
    157             bb = self.blkmap[offset]
    158             self.builder.position_at_end(bb)
--> 159             self.lower_block(block)
    160 
    161         self.post_lower()

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/lowering.py in lower_block(self, block)
    175             except Exception as e:
    176                 msg = "Internal error:\n%s: %s" % (type(e).__name__, e)
--> 177                 raise LoweringError(msg, inst.loc)
    178 
    179     def create_cpython_wrapper(self, release_gil=False):

LoweringError: Failed at nopython (nopython mode backend)
Internal error:
NotImplementedError: Don't know how to allocate array with layout 'A'.
File "<ipython-input-6-ecc34e250d52>", line 4
```

I also tried copying the slices, which gave a pretty cryptic error:
```Python
In [6]: @numba.jit(nopython=True)
   ...: def g():
   ...:     x = np.arange(5)
   ...:     return x[1:].copy(order="C") - x[:-1].copy(order="C")
   ...: 

In [7]: g()
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/typeinfer.py in propagate(self, context, typevars)
     97             try:
---> 98                 constrain(context, typevars)
     99             except TypingError:

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/typeinfer.py in __call__(self, context, typevars)
    217         fnty = typevars[self.func].getone()
--> 218         self.resolve(context, typevars, fnty)
    219 

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/typeinfer.py in resolve(self, context, typevars, fnty)
    239             kw_args = dict(zip(kwds, args[n_pos_args:]))
--> 240             sig = context.resolve_function_type(fnty, pos_args, kw_args)
    241             if sig is None:

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/typing/context.py in resolve_function_type(self, func, args, kws)
     84         if isinstance(func, types.Callable):
---> 85             return func.get_call_type(self, args, kws)
     86 

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/types.py in get_call_type(self, context, args, kws)
    194     def get_call_type(self, context, args, kws):
--> 195         return self.template(context).apply(args, kws)
    196 

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/typing/templates.py in apply(self, args, kws)
    141         generic = getattr(self, "generic")
--> 142         sig = generic(args, kws)
    143 

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/typing/templates.py in generic(_, args, kws)
    268                 def generic(_, args, kws):
--> 269                     sig = method_resolver(self, ty, args, kws)
    270                     if sig is not None:

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/typing/builtins.py in resolve_copy(self, ary, args, kws)
    591         assert not args
--> 592         assert not kws
    593         retty = ary.copy(layout="C")

AssertionError: 

During handling of the above exception, another exception occurred:

TypingError                               Traceback (most recent call last)
<ipython-input-7-d65ffd94a45c> in <module>()
----> 1 g()

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    157         assert not kws
    158         sig = tuple([self.typeof_pyval(a) for a in args])
--> 159         return self.compile(sig)
    160 
    161     def inspect_llvm(self, signature=None):

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/dispatcher.py in compile(self, sig)
    320                                           self.py_func,
    321                                           args=args, return_type=return_type,
--> 322                                           flags=flags, locals=self.locals)
    323 
    324             # Check typing error if object mode is used

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in compile_extra(typingctx, targetctx, func, args, return_type, flags, locals, library)
    592     pipeline = Pipeline(typingctx, targetctx, library,
    593                         args, return_type, flags, locals)
--> 594     return pipeline.compile_extra(func)
    595 
    596 

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in compile_extra(self, func)
    315                 raise e
    316 
--> 317         return self.compile_bytecode(bc, func_attr=self.func_attr)
    318 
    319     def compile_bytecode(self, bc, lifted=(), lifted_from=None,

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in compile_bytecode(self, bc, lifted, lifted_from, func_attr)
    324         self.lifted_from = lifted_from
    325         self.func_attr = func_attr
--> 326         return self._compile_bytecode()
    327 
    328     def compile_internal(self, bc, func_attr=DEFAULT_FUNCTION_ATTRIBUTES):

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in _compile_bytecode(self)
    579 
    580         pm.finalize()
--> 581         return pm.run(self.status)
    582 
    583 

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in run(self, status)
    207                     # No more fallback pipelines?
    208                     if is_final_pipeline:
--> 209                         raise patched_exception
    210                     # Go to next fallback pipeline
    211                     else:

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in run(self, status)
    199             for stage, stage_name in self.pipeline_stages[pipeline_name]:
    200                 try:
--> 201                     res = stage()
    202                 except _EarlyPipelineCompletion as e:
    203                     return e.result

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in stage_nopython_frontend(self)
    412                 self.args,
    413                 self.return_type,
--> 414                 self.locals)
    415 
    416         with self.fallback_context('Function "%s" has invalid return type'

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/compiler.py in type_inference_stage(typingctx, interp, args, return_type, locals)
    707 
    708     infer.build_constrain()
--> 709     infer.propagate()
    710     typemap, restype, calltypes = infer.unify()
    711 

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/typeinfer.py in propagate(self)
    397                 print("propagate".center(80, '-'))
    398             oldtoken = newtoken
--> 399             self.constrains.propagate(self.context, self.typevars)
    400             newtoken = self.get_state_token()
    401             if config.DEBUG:

/home/alan/miniconda3/envs/test/lib/python3.4/site-packages/numba/typeinfer.py in propagate(self, context, typevars)
    102                 msg = "Internal error at {con}:\n{err}"
    103                 raise TypingError(msg.format(con=constrain, err=e),
--> 104                                   loc=constrain.loc)
    105 
    106 

TypingError: Failed at nopython (nopython frontend)
Internal error at <numba.typeinfer.CallConstrain object at 0x7fc015f8a198>:

File "<ipython-input-6-4c558e2b1dd9>", line 4
```

**Comment by pitrou** [comment]
This is definitely a bug in Numba. Perhaps @jriehl wants to take a look when he's back :)

**Comment by pitrou** [comment]
Sorry for the delay on this. The snippet now works correctly, so I'm closing this issue.

**Comment by InonS** [comment]
Has this been merged? what release would I see it?

**Comment by seibert** [comment]
Given the time of the comment, I would assume this was fixed sometime before Numba 0.22.

**Comment by InonS** [comment]
@seibert Cool, thanks.
